src/Gaffer/MetadataAlgo.cpp

```diff
@@ -177,6 +178,28 @@ bool affectedByChange( const Node *node, IECore::TypeId changedNodeTypeId, const
 	return node->isInstanceOf( changedNodeTypeId );
 }
 
+void copy( const GraphComponent *from, GraphComponent *to, const StringAlgo::MatchPattern &exclude, bool persistentOnly, bool persistent )
```

Comment:
I put a fair bit of thought into the `promote()` method so that it _hopefully_ covers both Ivan's and Daniel's use cases while also fitting Gaffer's existing needs. @ivanimanishi, I made the following deviations from what you had :

- Make `parent` argument optional, since for the common case of promoting directly to the node above specifying it is redundant.
- Replace the `copyMetadata` boolean with an `excludeMetadata` match pattern. This allows us to accommodate the Box case where we want to copy everything except the layout metadata, and also fits a use case of @danieldresser's I saw where he excluded something subtly different.
- Register the metadata as nonpersistent, since if the plugs aren't dynamic it doesn't need to be (this matches @danieldresser's version of the function).
- Omit the `name` argument, since it was rarely used, and it's trivial to rename the plug after promotion.

Cheers...
John